File: Python/src/advanced/mt_data_race.py

import threading
import time
import random

# Liste globali condivise, funziona anche con variabili
safe_list = []
unsafe_list = []
# Lock per proteggere la lista sicura
safe_list_lock = threading.Lock()

def unsafe_increment(increment):
  global unsafe_list
  if unsafe_list:
    last = unsafe_list[-1]
    time.sleep(random.uniform(0.0001, 0.001))  # Ritardo artificiale
    unsafe_list.append(last + increment)
  else:
    time.sleep(random.uniform(0.0001, 0.001))  # Ritardo artificiale
    unsafe_list.append(increment)

def safe_increment(increment):
  global safe_list
  with safe_list_lock:
    if safe_list:
      last = safe_list[-1]
      time.sleep(random.uniform(0.0001, 0.001))  # Ritardo artificiale
      safe_list.append(last + increment)
    else:
      time.sleep(random.uniform(0.0001, 0.001))  # Ritardo artificiale
      safe_list.append(increment)

def data_race_example(verbose=False):
  global safe_list, unsafe_list
  safe_list = []  # Svuota liste
  unsafe_list = []  # Svuota liste
  num_threads = 100  # Numero di thread
  threads = []

  #  parte unsafe
  for i in range(0,num_threads): # == (num_threads)
    increment = random.randint(-5, 5)  # Incremento casuale
    thread0 = threading.Thread(target=unsafe_increment, args=(increment,))
    thread1 = threading.Thread(target=safe_increment, args=(increment,))
    threads.append(thread0)
    threads.append(thread1)
    thread0.start()
    thread1.start()


  for thread in threads:
    thread.join()

  if verbose:
    print(f"Valore finale del contatore sicuro: {sum(safe_list)}")
    print(f"Valore finale del contatore non sicuro: {sum(unsafe_list)}")
    print(f"I due contatori sono uguali? {sum(safe_list) == sum(unsafe_list)}")
    
  return sum(safe_list) == sum(unsafe_list)


def grandi_numeri_mt_rc():
  tot = 0
  tgn = 100
  for _ in range(0,tgn):
    if not data_race_example():
      tot += 1
  print(f"I due contatori sono risultati diversi {tot} volte su {tgn}")

class TestDataRace:
  @classmethod
  def run(cls):
    grandi_numeri_mt_rc()

File: c/src/advanced/mt_data_race.c
#include "mt_data_race.h"

int safe_counter = 0;
int unsafe_counter = 0;

pthread_mutex_t safe_counter_mutex = PTHREAD_MUTEX_INITIALIZER;

static void* unsafe_increment(void* arg) {
  int increment = *((int*)arg);
  sleep(rand() % 3 * 0.001);  // delay artificiale
  unsafe_counter += increment;
  return NULL;
}

static void* safe_increment(void* arg) {
  int increment = *((int*)arg);
  sleep(rand() % 3 * 0.001);  // delay artificiale
  pthread_mutex_lock(&safe_counter_mutex);
  safe_counter += increment;
  pthread_mutex_unlock(&safe_counter_mutex);
  return NULL;
}

void data_race_example(int verbose) {
  safe_counter = 0;  // reset il contatore sicuro
  unsafe_counter = 0;  // reset il contatore non sicuro

  int num_threads = 100;  // numero di thread
  pthread_t threads[num_threads * 2];
  int increments[num_threads * 2];

  srand(time(NULL));

  // crea e avvia i thread
  for (int i = 0; i < num_threads; i++) {
    increments[i] = rand() % 11 - 5;  // incremento casuale tra -5 e 5 unsafe
    pthread_create(&threads[i], NULL, unsafe_increment, &increments[i]);

    increments[num_threads + i] = rand() % 11 - 5;  // incremento casuale tra -5 e 5 safe
    pthread_create(&threads[num_threads + i], NULL, safe_increment, &increments[num_threads + i]);
  }

  for (int i = 0; i < num_threads * 2; i++)
    pthread_join(threads[i], NULL);

  if (verbose) {
    printf("Valore finale del contatore sicuro: %d\n", safe_counter);
    printf("Valore finale del contatore non sicuro: %d\n", unsafe_counter);
    printf("I due contatori sono uguali? %s\n", safe_counter == unsafe_counter ? "Sì" : "No");
  }
}


void grandi_numeri_mt_rc() {
  int tot = 0;
  int tgn = 100;
  for (int i = 0; i < tgn; i++) {
    data_race_example(0);
    if (safe_counter != unsafe_counter) 
      tot++;
  }
  printf("I due contatori sono risultati diversi %d volte su %d\n", tot, tgn);
}
